Evaluation Bug #2: Improve the error handling features in the UI.

soumyo123-prog opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following:
- Improves the UX in case the backend api service fails to fetch the version history data for a particular exploration at a particular version.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

I have stimulated an error during fetching a particular version history in my local system. Here, as the error was raised in the backend, the modal was immediately closed and error message was shown to the user:


https://user-images.githubusercontent.com/70739349/204028635-af81c571-1b3b-4d73-8875-ad24ce4be44e.mp4

For narrow screens:


https://user-images.githubusercontent.com/70739349/204044117-3f414764-327f-4fba-bfeb-af6d6bf3f890.mp4



When the backend API fails during the initial fetching, the UI would look like the following:

![Screenshot from 2022-11-25 22-18-18](https://user-images.githubusercontent.com/70739349/204028659-9e85012d-252d-426f-911b-6e1e6144b864.png)


#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

soumyo123-prog committed:
improve the error handling UI

seanlip commented:
Hi @soumyo123-prog, thanks for the PR. I don't think this counts as handling the error gracefully -- graceful handling of errors (especially cases that are expected to occur in practice) should not be handled with "orange alerts".

I would suggest that, when an error like this occurs, show red text next to the corresponding arrow button (along the lines of "Unable to display version history for version #NUM, due to an error loading that version of that exploration.") and disable that arrow button. Try to make sure it looks nice and properly aligned. You can make the red text a bit smaller, similar to e.g. the red text that shows up in the signup page if you don't enter a valid username.

soumyo123-prog committed:
some improvements

soumyo123-prog commented:
@seanlip Pushed the changes. PTAL. Thanks! I think we can go ahead with the testing now parallel with resolving ui issues (if any).

seanlip commented:
Looks good! Just one thing: left-align the red text, and make it not extend all the way to the right side (maybe just to the center line or slightly over). That keeps it associated with the left button.

seanlip commented:
(I also agree with you that testing can happen in parallel.)

kevintab95 commented on core/templates/pages/exploration-editor-page/modal-templates/state-version-history-modal.component.html:
```diff
@@ -91,4 +94,10 @@ <h3>
     width: 4rem;
   }
+  .error-message {
+    color: red;
```
Please prefer hex value for colors, per [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).

soumyo123-prog replied:
Done.

kevintab95 commented on core/templates/pages/exploration-editor-page/modal-templates/state-version-history-modal.component.ts:
```diff
@@ -51,6 +51,7 @@ export class StateVersionHistoryModalComponent
   @Input() oldState!: State;
   @Input() newStateName!: string;
   @Input() oldStateName!: string;
+  validationError: boolean = false;
```
```suggestion
  validationErrorIsShown: boolean = false;
```

Ditto elsewhere in this file.

soumyo123-prog replied:
Done.

kevintab95 reviewed (commented):
Thanks! Just 2 nits from my end.

kevintab95 commented:
Tested on the server. It seems the disabling of the button and error message seems to work as described in the PR description. However, I do see some console error, should we handle those as well?

https://user-images.githubusercontent.com/11008603/203912185-96dec941-94e8-4a19-abc8-7c6d9709ce98.mp4

These are the two main errors I see:

1.
```
TypeError: Cannot read properties of null (reading 'lastEditedVersionNumber')
    at Er.<anonymous> (settings-tab.component.ts:358:12)
    at Generator.next (<anonymous>)
    at a (exploration_editor.69ce59d1c23a65e79033.bundle.js:4135:36793)
    at e.invoke (zone.js:386:30)
    at Object.onInvoke (core.js:28580:33)
    at e.invoke (zone.js:385:36)
    at t.run (zone.js:143:47)
    at zone.js:891:38
    at e.invokeTask (zone.js:421:35)
    at Object.onInvokeTask (core.js:28567:33)
    at _ (zone.js:832:39)
    at zone.js:739:21
    at a (exploration_editor.69ce59d1c23a65e79033.bundle.js:4135:36811)
    at e.invoke (zone.js:386:30)
    at Object.onInvoke (core.js:28580:33)
    at e.invoke (zone.js:385:36)
    at t.run (zone.js:143:47)
    at zone.js:891:38
    at e.invokeTask (zone.js:421:35)
    at Object.onInvokeTask (core.js:28567:33)
```

2.
```
TypeError: Cannot read properties of null (reading 'lastEditedVersionNumber')
    at exploration-editor-tab.component.ts:385:24
    at e.invoke (zone.js:386:30)
    at Object.onInvoke (core.js:28580:33)
    at e.invoke (zone.js:385:36)
    at t.run (zone.js:143:47)
    at zone.js:891:38
    at e.invokeTask (zone.js:421:35)
    at Object.onInvokeTask (core.js:28567:33)
    at e.invokeTask (zone.js:420:40)
    at t.runTask (zone.js:188:51)
    at _ (zone.js:832:39)
    at zone.js:898:21
    at e.invokeTask (zone.js:421:35)
    at Object.onInvokeTask (core.js:28567:33)
    at e.invokeTask (zone.js:420:40)
    at t.runTask (zone.js:188:51)
    at b (zone.js:601:39)
    at t.invokeTask [as invoke] (zone.js:507:25)
    at p (zone.js:1671:18)
    at XMLHttpRequest.h (zone.js:1708:25)
```

soumyo123-prog committed:
some fixes

soumyo123-prog commented:
> Tested on the server. It seems the disabling of the button and error message seems to work as described in the PR description. However, I do see some console error, should we handle those as well?
> 
>  console-errors-soumya.mp4 
> These are the two main errors I see:
> 
> 
> ```
> TypeError: Cannot read properties of null (reading 'lastEditedVersionNumber')
>     at Er.<anonymous> (settings-tab.component.ts:358:12)
>     at Generator.next (<anonymous>)
>     at a (exploration_editor.69ce59d1c23a65e79033.bundle.js:4135:36793)
>     at e.invoke (zone.js:386:30)
>     at Object.onInvoke (core.js:28580:33)
>     at e.invoke (zone.js:385:36)
>     at t.run (zone.js:143:47)
>     at zone.js:891:38
>     at e.invokeTask (zone.js:421:35)
>     at Object.onInvokeTask (core.js:28567:33)
>     at _ (zone.js:832:39)
>     at zone.js:739:21
>     at a (exploration_editor.69ce59d1c23a65e79033.bundle.js:4135:36811)
>     at e.invoke (zone.js:386:30)
>     at Object.onInvoke (core.js:28580:33)
>     at e.invoke (zone.js:385:36)
>     at t.run (zone.js:143:47)
>     at zone.js:891:38
>     at e.invokeTask (zone.js:421:35)
>     at Object.onInvokeTask (core.js:28567:33)
> ```
> 
> 
> ```
> TypeError: Cannot read properties of null (reading 'lastEditedVersionNumber')
>     at exploration-editor-tab.component.ts:385:24
>     at e.invoke (zone.js:386:30)
>     at Object.onInvoke (core.js:28580:33)
>     at e.invoke (zone.js:385:36)
>     at t.run (zone.js:143:47)
>     at zone.js:891:38
>     at e.invokeTask (zone.js:421:35)
>     at Object.onInvokeTask (core.js:28567:33)
>     at e.invokeTask (zone.js:420:40)
>     at t.runTask (zone.js:188:51)
>     at _ (zone.js:832:39)
>     at zone.js:898:21
>     at e.invokeTask (zone.js:421:35)
>     at Object.onInvokeTask (core.js:28567:33)
>     at e.invokeTask (zone.js:420:40)
>     at t.runTask (zone.js:188:51)
>     at b (zone.js:601:39)
>     at t.invokeTask [as invoke] (zone.js:507:25)
>     at p (zone.js:1671:18)
>     at XMLHttpRequest.h (zone.js:1708:25)
> ```

@kevintab95 I have added the fix now. I have applied similar fixes as I had applied in state-version-history-modal-component. These errors were coming because the fetching of version history from the exploration editor tab component did not handle errors earlier. I have also added a screenshot of how the UI will look like if the backend api fails during the initial fetching of the version history.

soumyo123-prog commented:
@seanlip PTAL at the new video and screenshot uploaded.

seanlip commented:
This looks good. I suggest making the width of the "red text area" a bit narrower in the modal, but in general it looks fine!

One question, could you also please show what it looks like on narrow screens?

soumyo123-prog commented:
> This looks good. I suggest making the width of the "red text area" a bit narrower in the modal, but in general it looks fine!
> 
> One question, could you also please show what it looks like on narrow screens?

Uploaded the video for the narrow screens. Also, this feature is meant for screen sizes greater than 768px.

seanlip reviewed (approved):
LGTM, thanks!

> Also, this feature is meant for screen sizes greater than 768px.

It is still possible for users to view this on narrow screens, so we have to handle that case too.

seanlip commented:
@gp201 @Lawful2002 PTAL ASAP as codeowners, this needs to get merged for GSoC. Thanks!

soumyo123-prog committed:
reduce width of the error message

soumyo123-prog committed:
fix some tests and other errors

gp201 commented on core/templates/pages/exploration-editor-page/editor-tab/exploration-editor-tab.component.ts:
```diff
@@ -380,12 +389,16 @@ export class ExplorationEditorTabComponent
-              response.lastEditedCommitterUsername
-            );
+          ).then((response) => {
+            if (response !== null) {
+              this.versionHistoryService.insertStateVersionHistoryData(
+                response.lastEditedVersionNumber,
+                response.stateInPreviousVersion,
+                response.lastEditedCommitterUsername
+              );
+            } else {
+              this.validationErrorIsShown = true;
+            }
```
Is there a test for if and else condition?

soumyo123-prog replied:
Added it now.

gp201 commented on core/templates/pages/exploration-editor-page/modal-templates/metadata-version-history-modal.component.spec.ts:
```diff
@@ -289,8 +289,8 @@ describe('Metadata version history modal', () => {
     expect(alertsService.addWarning).toHaveBeenCalledWith(
-      'Could not fetch the version history data. ' +
-      'Please reload the page and try again.'
+      'Version history of cannot be explored further ' +
```
Add a line to test for `ngbActiveModal`

soumyo123-prog replied:
I have changed this part of code along with the tests.

gp201 reviewed (approved):
Thanks, @soumyo123-prog ! I left some comments. PTAL

gp201 reviewed (changes requested):
Thanks, @soumyo123-prog ! I left some comments. PTAL

soumyo123-prog committed:
fixes for settings tab

soumyo123-prog committed:
fix some tests

soumyo123-prog commented:
@gp201 I have pushed the required changes. PTAL. Thanks!

gp201 reviewed (approved):
LGTM

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into fix_evaluation_bugs_2
